### ISSUE TITLE
Bad Performance.

### ISSUE THREAD

**Issue by AmbientMalice** [gonetz/GLideN64#470]
I'm just filing this as a general issue. This plugin runs really, really badly for me. I'm running a Core Quad @ 3Ghz, and an Nvidia GTX 750Ti. Emulator is Project 64 2.2, compiled from latest source.

Some games are acceptable. But games such as Perfect Dark have massive GFX\RSP usage spikes when using goggles\camspy\etc. The VIs dip below 60 even walking around Carrington Institute. Depth Buffer to RDRAM tends to be a performance killer, pushing games over the line.

Either this is some Nvidia or PC-specific problem, or this plugin needs optimisation. It runs Perfect Dark far worse than Glide64 did, and Glide64 ran it terribly, due to framebuffer overheads. Lots of games have sudden VI dips when the plugin chokes on something. Vigilante 8 is basically unplayable in HLE, much less LLE.

TL;DR EDIT:

Problem turns out to be Nvidia power management. Darn you, Nvidia and your wonky adaptive power management.

**Comment by purplemarshmallow** [comment]
I think the problem is framebuffer emulation. It runs really fast on my system with framebuffer emulation disabled. But only some games are affected like perfect dark. But depth buffer to RDRAM is slow in every game. I'm using Nvidia too.

**Comment by etking** [comment]
I can also confirm that the emulation of some games runs extremely slow in 1080p, smaller resolutions are fine. In Banjo-Kazooie as soon as the Carrot-like enemy in the starting area is visible, the frame rate drops and the game gets unplayable. (i72670QM,+ GT555M)

Perfect Dark however runs faster and better than in any other plug-in, even in 1080p.

**Comment by purplemarshmallow** [comment]
I tested Perkect Dark Carrington Institute. I'm getting 200-500 VI/s without framebuffer emulation and 65-250 VI/s with framebuffer emulation. So framebuffer emulation is very demanding but it's still smooth on my system.

**Comment by purplemarshmallow** [comment]
But I don't have any problems in Banjo-Kazooie

**Comment by thecoreyburton** [comment]
I have no issues with any of the above games on my rig, the slowest I think it dropped was to about 55VIs during graphically intensive parts. FSAA, Frame Buffer and texture filtering effects will each take a hit to your performance depending on settings and what's going on in the game at the time. This is just how graphic processing works.

If performance is an issue, it's not a plugin problem but rather hardware related. You could use an alternative, less intensive plugin or upgrade your hardware. VGA cards are always coming down in price and if money's a factor an AMD R9-290 should give decent performance for relatively cheap prices compared to the Nvidia counterparts. If you can afford to fork out more, I'd stick with Nvidia though.

**Comment by AmbientMalice** [comment]
@thecoreyburton I doubt the issue is GPU related. (Unless it is an Nvidia problem.) The GTX 750Ti is a 2014 GPU, among the first of Nvidia's Maxwell line.


>@purplemarshmallow
>I tested Perkect Dark Carrington Institute. I'm getting 200-500 VI/s without framebuffer emulation and 65-250 VI/s with framebuffer emulation. So framebuffer emulation is very demanding but it's still smooth on my system.

That is not good. The Carrington Insitute should not be that resource intensive. The moment you add night vision goggles or a similar framebuffer effect, the framerate is going to crash below 60. Try Deep Sea with infra red goggles.

Also, changing from Low Resolution mode to High Resolution mode using ingame settings hurts the framerate even more.

![low](https://cloud.githubusercontent.com/assets/10704701/7441213/d4472f6c-f120-11e4-9739-0d42748e61c6.png)

![high](https://cloud.githubusercontent.com/assets/10704701/7441217/f3a144d8-f120-11e4-8bcd-98c840c5b70f.png)

This is odd behavior for a 3Ghz+ Quad Core CPU and a modern Nvidia GPU.

**Comment by purplemarshmallow** [comment]
35 VI/s with infra red goggles :(

**Comment by gonetz** [comment]
I have save with infra red only for 1964. Works fine. Visually >60 VI/S without speed limit. Can't say how much, because 1964 VI counter works wrong. Glide64 is faster with no doubt. It is generally 2-3  times faster than GLideN64, and 2 times faster than glN64.

**Comment by AmbientMalice** [comment]
>I have save with infra red only for 1964. Works fine. Visually >60 VI/S without speed limit. Can't say how much, because 1964 VI counter works wrong. Glide64 is faster with no doubt. It is generally 2-3 times faster than GLideN64, and 2 times faster than glN64. 

I get 30-60VIs in carrington instutute in 1964. This jumps to 60-70 with FB disabled.

Is it possible this unusual overhead could be Nvidia-related? You're running AMD, right?

**Comment by gonetz** [comment]
I use pretty old AMD Radeon HD 6800 Series card.
I can't say, it is NVidia related. I have performance issues in some places too.
I'd like to get good NVidia card because NVidia has good set of development tools.
Can't afford it right now.

**Comment by WizLizard** [comment]
It seems to be related to the public release. When running yoshi's story with early release i get constant full speed but with the public release i get huge dips to 40 vi. Both early release and public release are on max settings. Although i did update my nvidia drivers before public release. I also get slowdown in mario kart 64 in public release but not in early release? 

i7 4800mq @3.8 ghrz  Nvidia gtx 880m

**Comment by AmbientMalice** [comment]
Good news, @gonetz! I've discovered the culprit for my awful framerates, especially at higher resolutions, is Nvidia's power management. Now the bad news is that I have no clue how to fix this beyond manually setting power management to "Prefer Maximum Performance". This unfortunately makes things a smidge less user friendly. (Dolphin is\was affected by this same problem.)

**Comment by WizLizard** [comment]
@AmbientMalice  This isn't the case for me. I have my global settings set to prefer maximum performance in my Nvidia control panel and i still get bad performance in some games. Even if i add project64 to the program list within the control panel. I believe something is going on with the texture enhancement's. There seems to be no difference in my cpu load between any of the enhancements. x2 puts the cpu load at 25% just as 5xbrz does in certain areas in games like yoshi's story. For example the white boxes that give you tips on game mechanics etc. Sure the 5xbrz looks far better than the x2 but performance seems identical which is bizarre. Even with single smooth filtering 1 i get the same poor performance? With all enhancements off it's smooth as butter. The x2 should not be as much of a performance hit as the 5xbrz but both alone cpu load reaches 25-35% and VI's drop down as far as 40-30. Check out this video i made demonstrating what i'm describing. 
https://www.youtube.com/watch?v=jnrKp9Ecr24

System Specs:
PC: MSI GT70 2PE Dominator Pro
Operating System: Windows 8.1 64-bit
Processor/CPU: 4th Gen Intel Core i7 4800MQ 2.8GHz-3.8GHz(Turbo)
Video Card/GPU: Nvidia GTX 880m 8gb GDDR5
Memory/RAM: 32gb GDDR3

**Comment by AmbientMalice** [comment]
@WizLizard My impression is everything texture related is a bit janky at the moment, but I haven't really done any testing there.

**Comment by WizLizard** [comment]
@AmbientMalice all i know is i wasn's having these issues with early release even with enhancements and filtering maxed simultaneously.

**Comment by CraigClements** [comment]
I'm using a Core i5 4670K 3.4GHz (stock) with GeForce GTX 670, and do not experience issues with the Camspy, IR Scanner, Nightvision or even the Slayer Fly-by-Wire Rocket secondary function.

However, when running around the Carrington Institute with GLideN64, I will sometimes see a small dip in framerate to about 54 V/Is on a fairly rare occasion. Upon further investigation with the frame limiter off, it appears that wandering around the Carrington Institute - I get between 60-62 V/Is most of the time. Seems my hardware is just barely enough for this. This is without sound syncing. Similarly with Glide64 - I get around 60-110 V/Is depending on where I am standing.

**Comment by AmbientMalice** [comment]
@CraigClements Does changing to Prefer Maximum Performance change anything?

**Comment by CraigClements** [comment]
I can test that tomorrow, I'm still not sure where exactly the bottleneck is what is causing the lag. My GPU is hardly registering use, I know that much. But you might be right there. Obviously most other games are running much faster than PD. PD is a pretty good benchmark for a high requirement game though. I'm using HLE mode as well.

**Comment by AmbientMalice** [comment]
>@CraigClements
>My GPU is hardly registering use, I know that much.

My GTX 750Ti registers 30% load when in adaptive mode. PD suffers massive slowdowns to ~40VIs despite never going above 30% load. In Performance Mode, the load stays between 7-12%, with only occasional dips to 58-59VIs.

GPU load is not a reliable measure because the problem is that the GPU is underclocking itself because the drivers are misdetecting the emulator's load.

**Comment by CraigClements** [comment]
I retested the Carrington Institute with GLideN64 with both of the settings. There doesn't seem to be a big difference in terms of V/Is for me with the limiter set to off, but obviously I am only testing one area and in one game on one set of hardware, so who knows what can happen on other configs. It might help some users better than others.

However my CPU load darts around 21-27% with the frame limiter in PJ64 set to on. This suggests that the plugin is close to (or reaches) max utilisation of one of my CPU cores in this usage case. Makes sense as neither PJ64 (or presumably GLideN64) can make use of the extra CPU threads to boost general speed. It would explain the slowdown I get from time to time. I suspect that someone using a higher clocked CPU than mine probably wouldn't experience any lag at all with the plugin when wandering around the Carrington Institute. I'm not sure about the influence of the graphics card from what I can see, but I am guessing that once the card is faster than a certain point, very little can be achieved though upgrading or overclocking it. Graphics cards below that point would otherwise benefit. But this is just a guess.

**Comment by Papermanzero** [comment]
Issue is still present. I have some stutters during yoshis story or majoras mask. It is playable but not completely free of stutters.

**Comment by oddMLan** [comment]
GLideN64 needs to set a high resolution timer on Windows to force drivers to run in high performance profile instead of going down to Power-saving mode. Maybe that would solve the stuttering for some people. 

I forgot the name of the issue this very point was addressed before. If somebody knows please reference it here.

**Comment by Papermanzero** [comment]
I guess you mean the Power Saving Mode from nvidia. 
I thought about that already. However the stutters are still present if you fully deactivate the power saving mode (even after the shader cache is built).

**Comment by fzurita** [comment]
Here it is: https://github.com/gonetz/GLideN64/issues/1346

**Comment by Papermanzero** [comment]
Any proposals how to boost the performance via settings. I set up mupen64plus from scratch, however the performance of the newest WIP2 build is really bad. Energy settings for cpu and gpu are deactivated. But still the performance is not getting better.

**Comment by fzurita** [comment]
Did you enable high resolution timers?

**Comment by Papermanzero** [comment]
It is better, but does not solve the problem.

**Comment by fzurita** [comment]
What GPU do you have? Can you try disabling color buffer to RDRAM?

**Comment by Papermanzero** [comment]
I have a GTX760 overclocked. CPU is an i7 4790.
The color buffer is disabled by default. I am using the .default settings from GlideN64.

**Comment by loganmc10** [comment]
What change are you referring to?

**Comment by Papermanzero** [comment]
This one: 
https://github.com/mupen64plus/mupen64plus-core/commit/cc8226383993f9f509ba36ac463d7acad7b339ba

I will try one of the last mupen builds to see if it is related to this change.

**Comment by loganmc10** [comment]
Oh that's not going to affect Majora's Mask, and it only does anything if you're using rsp-cxd4, for Majora's Mask you should use rsp-hle

**Comment by Papermanzero** [comment]
Do you have any idea, how to increase mupen64plus performance?

**Comment by fzurita** [comment]
Something feels odd, your configuration seems perfect for running GLideN64. How is your performance with Glide64mk2? Can you post your mupen64plus.cfg?

**Comment by loganmc10** [comment]
Can you try Project64 with GLideN64? That'll tell us if your issue is with GLideN64 or mupen64plus

**Comment by Papermanzero** [comment]
So it is definitely mupen. Glide64mk2 shows the same fps issues. project64 is also running well, although it has other issues and I don't like the emulator. ^^

**Comment by fzurita** [comment]
Ok, it's probably the speed limiter in audio-sdl plugin.

**Comment by Papermanzero** [comment]
I searched already for that. However my settings are:

```
[Audio-SDL]

# Mupen64Plus SDL Audio Plugin config parameter version number
Version = 1
# Frequency which is used if rom doesn't want to change it
DEFAULT_FREQUENCY = 32767
# Swaps left and right channels
SWAP_CHANNELS = False
# Size of primary buffer in output samples. This is where audio is loaded after it's extracted from n64's memory.
PRIMARY_BUFFER_SIZE = 16384
# Fullness level target for Primary audio buffer, in equivalent output samples. This value must be larger than the SECONDARY_BUFFER_SIZE. Decreasing this value will reduce audio latency but requires a faster PC to avoid choppiness. Increasing this will increase audio latency but reduce the chance of drop-outs.
PRIMARY_BUFFER_TARGET = 2048
# Size of secondary buffer in output samples. This is SDL's hardware buffer. The SDL documentation states that this should be a power of two between 512 and 8192.
SECONDARY_BUFFER_SIZE = 1024
# Audio resampling algorithm. src-sinc-best-quality, src-sinc-medium-quality, src-sinc-fastest, src-zero-order-hold, src-linear, speex-fixed-{10-0}, trivial
RESAMPLE = "src-sinc-best-quality"
# Volume control type: 1 = SDL (only affects Mupen64Plus output)  2 = OSS mixer (adjusts master PC volume)
VOLUME_CONTROL_TYPE = 1
# Percentage change each time the volume is increased or decreased
VOLUME_ADJUST = 5
# Default volume when a game is started.  Only used if VOLUME_CONTROL_TYPE is 1
VOLUME_DEFAULT = 80
```

**Comment by fzurita** [comment]
Can you try not setting the audio plugin? I think it should fall back to the dummy audio plugin. See if performance improves.

**Comment by Papermanzero** [comment]
I tried. Mupen is not starting with the statement no audio plugin configured.

**Comment by fzurita** [comment]
Sorry, for the audio plugin put in "dummy" instead SDL.

**Comment by loganmc10** [comment]
@Papermanzero can you try using my fork of mupen64plus-audio-sdl

https://github.com/loganmc10/mupen64plus-audio-sdl

There is a branch called "sync". If you build that, Audio-SDL will have a new configuration option called "AUDIO_SYNC", which should be disabled.

Or like @fzurita mentioned, you can set the audio plugin to "dummy", or delete the mupen64plus-audio-sdl.dll file

**Comment by Papermanzero** [comment]
Thanks a lot @loganmc10 
I will try today if the plugin is working.
As an alternative, I can upload my version and configuration. 
I am using m64py which seems does not provide the functionality to select the dummy plugin or to disable a plugin.

**Comment by loganmc10** [comment]
@Papermanzero here is a win64 build of audio-sdl that has the new configuration option to disable AV sync:

https://drive.google.com/open?id=0B8iAuv3fmhb_NW0wYmVpR1VNRkk

Also, you can try this:
http://m64p.s3.amazonaws.com/m64p/mupen64plus-GLideN64-win64-201705.zip

That is a win64 build of the latest mupen64plus + GLideN64, you should be able to just unzip that and play (you run "mupen64plus-gui.exe"), that GUI also has the option to select the "dummy" audio plugin.

**Comment by Papermanzero** [comment]
So I tested all configurations.
The audio plugin is unfortunately not working with m64py.
The warning pop up appears with "Failed to execute script m64py".

However I tested the audio plugin and the behavior with m64p.
I can confirm it is related to the audio plugin.

Majoras Mask should have 17 FPS in the PAL and 20 FPS in the NTSC Version.
With the old Audio Plugin I have 15 FPS(sometimes 13 FPS) in the PAL and 18FPS (sometimes 15 FPS) in the NTSC.
With the new Audio Plugin (Sync deactivated) it jumps between 15 FPS and 17 FPS in the PAL (seems ok). In the NTSC bewtween 18FPS and 20 FPS (seems also ok).

**Comment by loganmc10** [comment]
I have a PR open with those changes:

https://github.com/mupen64plus/mupen64plus-audio-sdl/pull/16

If you could post your findings there it would be helpful, hopefully that will give them some motivation to have it merged.

**Comment by fzurita** [comment]
The way audio SDL is implemented is buggy. The SLES audio plugin used to be implemented in a similar way, but I redesigned the timing of it so it could run on a separate thread. The same changes could be applied to the SDL plugin and it would probably work better.

Before I made my changes to the SLES audio plugin, Android users also reported many similar issues.

**Comment by olivieryuyu** [comment]
@AmbientMalice @loganmc10 

Can it be closed? I am sure that this issue is still running and if it is due to GlideN64.

**Comment by loganmc10** [comment]
It can be closed as far as I'm concerned, but @AmbientMalice opened the issue. I think there might still be some issues related to MSAA at high res, and I think we discussed removing the CacheSize parameter with @gonetz , but perhaps those can be separate issues
